chatcli_gpt/plugins.py

```diff
@@ -31,33 +39,60 @@ def extract_blocks(response_text, plugin):
     return matches
 
 
+def exec_bash(code):
+    try:
+        result = subprocess.run(
+            ["/bin/bash", "-c", code], capture_output=True, text=True
+        )
+    except Exception:  # pylint: disable=broad-except
+        print(traceback.format_exc())
+
+    return {
+        "result": result.stdout.strip(),
+        "error": result.stderr.strip(),
+    }
+
+
 def exec_python(code):
-    buffer = io.StringIO()
-
-    with contextlib.redirect_stdout(buffer):
-        global_scope = globals()
-        try:
-            mod = ast.parse(code, mode="exec")
-            if isinstance(mod.body[-1], ast.Expr):
-                last_expr = mod.body.pop()
-                exec(compile(mod, "<ast>", "exec"), global_scope)
-                result = eval(compile(ast.Expression(last_expr.value), "<ast>", "eval"), global_scope)
-                if result is not None:
-                    print(result)
-            else:
-                exec(code, global_scope)
-        except Exception:  # pylint: disable=broad-except
-            print(traceback.format_exc())
-
-    return buffer.getvalue().strip()
+    stdout = io.StringIO()
+    stderr = io.StringIO()
+
+    with contextlib.redirect_stdout(stdout):
+        with contextlib.redirect_stderr(stderr):
+            global_scope = globals()
+            try:
+                mod = ast.parse(code, mode="exec")
+                if isinstance(mod.body[-1], ast.Expr):
+                    last_expr = mod.body.pop()
+                    exec(compile(mod, "<ast>", "exec"), global_scope)
+                    result = eval(
+                        compile(ast.Expression(last_expr.value), "<ast>", "eval"),
+                        global_scope,
+                    )
+                    if result is not None:
+                        print(result)
+                else:
+                    exec(code, global_scope)
+            except Exception:  # pylint: disable=broad-except
+                print(traceback.format_exc())
```

Comment:
Need to add file=stderr if we're capturing separately.

Reply:
There are two levels of output:

1.  Output of the tool (search, pyrepl, shell), that output should be piped to GPT
2. Output of this program, this is excluded of the message sent to GPT.

With this leveling, Ithe traceback should only be printed instead of collected and sent back.

What's your thought?

(Could also address in a future PR).

Reply:
Agreed. These exceptions are the output of from the python code being executed, so I expect them to go back to GPT.